doc/rst/technotes/static.rst

```diff
@@ -0,0 +1,192 @@
+       var toReturn: c_array(real, tableSize);
+
+       toReturn[0] = seed1;
+       toReturn[1] = seed2;
+       for i in 2..<tableSize {
+           toReturn[i] = toReturn[i-1] + toReturn[i-2];
+       }
+
+       return toReturn;
+   }
+
+   proc getNthElement(x: int): real {
+       @functionStatic
+       ref table = computeExpensiveTable(1, 1);
+       return table[x];
+   }
+
+   writeln("Element 0: ", getNthElement(0));
+   writeln("Element 1: ", getNthElement(1));
+   writeln("Element 100: ", getNthElement(100));
+
+The output of the above program is as follows:
+
+.. code-block:: none
+
+   Computing an expensive table
+   Element 0: 1
+   Element 1: 1
+   Element 100: 5.73148e+20
+
+Importantly, the 'computing' line is only printed once, despite the three
+invocation of ``getNthElement``. The static variable in the above snippet is
+``table``. It's marked static using the ``@functionStatic`` annotation.
+Because of this annotation, the initialization expression of ``table``
+is only executed during the first call to ``getNthElement``.
+
+One important aspect of the snippet above is that the variable was declared
+using the ``ref`` intent. The reason for this is that ``table`` is
+assigned from the (stored) cached result. Thus, if the ``var`` or ``const``
+intents were used, each invocation of the function would result in copying
+the table from the memoized result to the local variable. If the table
+is large enough to warrant memoization, the copy could be expensive.
+
+Generic Functions
+-----------------
+In generic functions with static variables, one instance of the static
+variable is created for each instantiation. Consider the following example:
+
+.. code-block:: chapel
+
+   proc defaultValueForType(type x: numeric): x do return 0;
+   proc defaultValueForType(type x: string): x do return "";
+
+   proc genericFunction(arg) {
+       @functionStatic
+       ref acc = defaultValueForType(arg.type);
+
+       acc += arg;
+       return acc;
+   }
+
+   writeln(genericFunction(1));
+   writeln(genericFunction(2));
+   writeln(genericFunction(0.5));
+   writeln(genericFunction(0.25));
+   writeln(genericFunction("hello"));
+   writeln(genericFunction(" world"));
+
+In this example, one copy of the accumulator is created for each of the
+functions instantiations (``int``, ``real`` and ``string``). As a result,
+the output of the program is as follows:
+
+.. code-block:: none
+
+   1
+   3
+   0.5
+   0.75
+   hello
+   hello world
+
+Synchronization
+---------------
+
+The initialization of function-static variables is implicitly synchronized
+using of Chapel's ``atomic`` types. As a consequence, it's safe to call a
+function with static variables from multiple concurrent threads, as well as
+from several locales. In the latter case, the variable is stored on the first
+locale to initialize it ; support for alternative ways of sharing the data
+(e.g., replicating the precomputed data to all locales) is considered future
```

Comment:
you mention future work here, but then don't repeat this at the end in the bulleted list